### ISSUE TITLE
[rush] Is Bun support planned?

### ISSUE THREAD

**Issue by EdwardDrapkin** [microsoft/rushstack#4457]
Just wondering if there was a plan to add bun's package manager as supported to rush alongside pnpm/npm/yarn?

Thanks for your time!

**Comment by octogonz** [comment]
Probably not at this time.  It seems that Bun implements the [flawed NPM installation model](https://lfx.rushstack.io/pages/concepts/install_models/).  

For example, I tried converting the [demo/sbs-3](https://lfx.rushstack.io/pages/scenarios/side-by-side_versions/#step-3-side-by-side-version-trouble) branch to use a Bun workspace.  Its dependency graph is supposed to look like this:

<img src="https://raw.githubusercontent.com/microsoft/lockfile-explorer-demos/demo/sbs-3/common/images/lfx-demo-sbs-3.svg" />

Running `bun install` produces:

- **doppelgangers:** Duplicate copies of `p@1.2.3` under `b` and `c` which PNPM would avoid using symlinks.  Having two copies of the same package folder causes trouble such as singletons not being single, TypeScript having to process duplicate .d.ts files, etc.
- **a phantom dependency:** The `p@2.0.3` installation gets "hoisted" to the top-level `node_modules` folder. This is bad because other projects in the monorepo can accidentally import it without declaring a dependency in their own **package.json** file.  PNPM avoids this problem by using symlinks.

These problems are tolerable in a small single-team repo (in fact much of the NPM community is entirely unaware of it), but it's too painful for a large scale monorepo, which is the focus of Rush. It's why the Rush community has largely abandoned the NPM+Rush configuration. If Bun is able to implement a more correct installation model, that would make `bun install` more attractive.

**Comment by octogonz** [comment]
Apparently Bun only supports Windows Subsystem for Linux (WSL) not actual Windows/NTFS. This is might also be a blocker for companies that ship device apps that build using Windows.  This is less common, although it's characteristic of large scale monorepos to eventually accumulate every possible requirement. 🫠

**Comment by octogonz** [comment]
Of course, if someone from the community wants to contribute this feature, we would certainly support that.👍

**Comment by octogonz** [comment]
Well, the NPM and Yarn implementations predated workspaces, so they create a synthetic common/temp/package.json that depends on the union of all package.json files, and install that, then Rush creates symlinks for each project into that folder. This approach has benefits of preventing some phantom dependencies and enabling Rush-specific features such as preferred versions. Your other option would be to follow the approach of the PNPM implementation which uses a proper PNPM workspace, without Rush creating any symlinks. There are pros/cons to each approach.

The main thing I will say is that the world of node_modules folders is much more complicated than people assume, particularly for production monorepos. If you are serious about contributing we should do a community call where we can discuss it in detail.

**Comment by Niek** [comment]
With the bun 1.x releases, this looks more and more interesting. Full native Windows support has been added as well. Hopefully it's something that will be considered.

**Comment by VasekProchazka** [comment]
There was some huge updates for Deno in version 2, but also Bun with version 1.2. Is there any updates on this topic?

**Comment by TheBit** [comment]
@octogonz, @EdwardDrapkin, hi, long-time Rush + pnpm user here, who is considering migrating to bun, but can't live without Rush and isolated installs :)

With [Bun v1.2.19](https://bun.com/blog/bun-v1.2.19) (July 19, 2025) introducing [bun install --linker isolated](https://bun.com/docs/pm/cli/install#isolated-installs), the ecosystem is clearly evolving. Alongside tools like Biome, typescript-go, and JSR, it feels like the landscape is shifting.


Is it time to revisit this issue? I’d love to see this get the green light!

**Comment by octogonz** [comment]
We would need to do a technical investigation to figure out what are all the work items. What is broken/different in Bun? Can the work be broken down into stages (e.g. some middle ground where Node.js or compatibility hacks are used in the beginning)? And what are the key advantages?

Would you be interested to work on this?
